glusterd2/commands/volumes/brick-replace.go

```diff
@@ -0,0 +1,222 @@
+package volumecommands
+
+import (
+	"net/http"
+
+	"github.com/gluster/glusterd2/glusterd2/brick"
+	"github.com/gluster/glusterd2/glusterd2/bricksplanner"
+	"github.com/gluster/glusterd2/glusterd2/gdctx"
+	"github.com/gluster/glusterd2/glusterd2/peer"
+	restutils "github.com/gluster/glusterd2/glusterd2/servers/rest/utils"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+	"github.com/gluster/glusterd2/pkg/api"
+	gderrors "github.com/gluster/glusterd2/pkg/errors"
+
+	"github.com/gorilla/mux"
+	"github.com/pborman/uuid"
+)
+
+func registerReplaceBrickStepFuncs() {
+	var sfs = []struct {
+		name string
+		sf   transaction.StepFunc
+	}{
+		{"brick-repalce.PrepareBrick", prepareBricks},
+		{"brick-replace.ReplaceVolinfo", replaceVolinfo},
+		{"brick-replace.StartBrick", startBrick},
+	}
+	for _, sf := range sfs {
+		transaction.RegisterStepFunc(sf.sf, sf.name)
+	}
+}
+
+func replaceBrickHandler(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+
+	volname := mux.Vars(r)["volname"]
+
+	// Get Volume Info
+	vol, err := volume.GetVolume(volname)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+
+	// Unmarshal ReplaceBrickReq
+	var req api.ReplaceBrickReq
+	if err := restutils.UnmarshalRequest(r, &req); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, gderrors.ErrJSONParsingFailed)
+		return
+	}
+
+	subVols := vol.Subvols
+	var peerIDs []uuid.UUID
+	var srcBrickInfo brick.Brickinfo
+	m := make(map[string]bool)
+	brickCount := 0
+	subVolIndex := 0
+	brickIndex := 0
+
+	for index := range subVols {
+		for i, brick := range subVols[index].Bricks {
+			brickCount++
+			// Get Source Brick Info
+			if brick.PeerID.String() == req.SrcPeerID && brick.Path == req.SrcBrickPath {
+				subVolIndex = index
+				brickIndex = i
+				srcBrickInfo = brick
+			}
+			// Avoid duplicate appends
+			if _, ok := m[brick.PeerID.String()]; !ok {
+				m[brick.PeerID.String()] = true
+				peerIDs = append(peerIDs, brick.PeerID)
+			}
+		}
+	}
+
+	zones := make(map[string]bool)
+	// Get all zones of bricks in volume
+	for _, peerID := range peerIDs {
+		peer, err := peer.GetPeer(peerID.String())
+		if err != nil {
+			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+			return
+		}
+		if _, ok := zones[peer.Metadata["_zone"]]; !ok {
+			zones[peer.Metadata["_zone"]] = true
+		}
+	}
+	if len(zones) <= 0 {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "No zones available")
+		return
+	}
+
+	allPeers, err := peer.GetPeers()
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+	// Find peers which has zone apart from the ones already being used in volume.
+	var allPeerIDs []uuid.UUID
+	var validPeers []*peer.Peer
+	for _, p := range allPeers {
```

Comment:
Instead of getting all Peers, just collect the zones of the current subvolume and volume. Include that list as `ExcludeZones`

```
excludeZones = []

for svIndex, sv := range volinfo.Subvols {
	if filter.SubvolZonesOverlap && srcBrick.subvolIndex != svIndex {
		continue
	}
	for _, b := range sv.Bricks {
	    p = getpeer(b.PeerID)
	    excludeZones = append(p.Zone)
	}
}
```

With this no change required in `GetAvailableVgs` and also this function will get simplified.

Note: Add `SubvolZonesOverlap` to device filter